Incorrect MPE for Long Chains of Discrete Factor Graphs

## Description
Hi team, thanks for the wonderful library!

I've run into an issue where I have a fairly large Discrete Factor Graph that I believe is running into an underflow type of error. I've copied a minimal toy example below that shows the same symptom. I'm not sure where to turn for documentation of python bindings or how to debug the optimization further. Thank you for the help!

## Steps to reproduce
Toy example is a hidden markov model with 3 possible states and 200 Observations. In this example, the transition matrix is mostly identity. The desired state outputs any number [0, 199] with equal probability. All other states output 200 with probability 1. I set each measurement equal to the observation index (so only [0,200] are observed and the desired state should be preferred always).

``` python
import gtsam
import numpy as np

import gtbook
from gtbook.discrete import Variables
VARIABLES = Variables()

def toy_problem(num_states, num_obs, desired_idx):
    states = list(range(num_states))
    X = VARIABLES.discrete_series("X", range(num_obs), states)
    Z = VARIABLES.discrete_series("Z", range(num_obs), range(num_obs+1))
    graph = gtsam.DiscreteFactorGraph()

    # Mostly identity transition matrix
    transitions = np.eye(num_states)
    
    # Needed otherwise mpe is always state 0?
    transitions += 0.1/(num_states)
    
    transition_cpt = []
    for i in range(0, num_states):
        transition_row = "/".join([str(x) for x in transitions[i]])
        transition_cpt.append(transition_row)
    transition_cpt = " ".join(transition_cpt)

    for i in reversed(range(1, num_obs)):
        transition_conditional = gtsam.DiscreteConditional(X[i], [X[i-1]], transition_cpt)
        graph.push_back(transition_conditional)

    # Contrived example such that the desired state gives measurements [0, num_obs) with equal probability
    #   but all other states always give measurement num_obs
    obs = np.zeros((num_states, num_obs+1))
    obs[:,-1] = 1
    obs[DESIRED_STATE,0: -1] = 1
    obs[DESIRED_STATE,-1] = 0
    obs_cpt_list = []
    for i in range(0, num_states):
        obs_row = "/".join([str(z) for z in obs[i]])
        obs_cpt_list.append(obs_row)
    obs_cpt = " ".join(obs_cpt_list)
    
    # Contrived example where each measurement is its own index
    for i in range(0, num_obs):
        obs_conditional = gtsam.DiscreteConditional(Z[i], [X[i]], obs_cpt)
        factor = obs_conditional.likelihood(i)
        graph.push_back(factor)
    
    mpe = graph.optimize()
    vals = [mpe[X[i][0]] for i in range(num_obs)]
    print("MPE Vals: ")
    print(vals)
    print("Indices where mpe does not equal desired state:")
    print(np.where(np.array(vals) != DESIRED_STATE))

    # Missing many outputs
    sum_product = graph.sumProduct()
    breakpoint()

if __name__ == "__main__":
    NUM_STATES = 3
    NUM_OBS = 200
    DESIRED_STATE = 1

    toy_problem(NUM_STATES, NUM_OBS, DESIRED_STATE)
```

Output:
```
MPE Vals: 
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Indices where mpe does not equal desired state:
(array([138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150,
       151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163,
       164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176,
       177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189,
       190, 191, 192, 193, 194, 195, 196, 197, 198, 199]),)

```

## Expected behavior

Expected behavior is that the maximum probability estimate is equal to DESIRED_STATE (1) for all hidden states. This does not hold true for hidden states 138 and above. Additionally, the graph resulting from `sum_product = graph.sumProduct()` is missing many values for these incorrect states.

## Environment

I'm running on Python 3.8.13, gtsam version 4.2a9, Ubuntu 20.04

Hi yes, this does look it like resolves my issue in both the test case and in my project!

When I try to run all of the unit tests with `make check`, it says that a different unit test now fails:
```
99% tests passed, 1 tests failed out of 286

Total Test time (real) =   3.99 sec

The following tests FAILED:
         82 - testDiscreteFactorGraph (Failed)
Errors while running CTest
make[3]: *** [CMakeFiles/check.dir/build.make:76: CMakeFiles/check] Error 8
make[2]: *** [CMakeFiles/Makefile2:2932: CMakeFiles/check.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2939: CMakeFiles/check.dir/rule] Error 2
make: *** [Makefile:229: check] Error 2
```

Fortunately, this might just require the expected values to be adjusted? The old expected values seem to be unnormalized versions of the new ones.
```
 82/286 Test  #82: testDiscreteFactorGraph .......................***Failed    0.01 sec
Not equal:
expected:
 f[ (1,2), (2,2), ]
 Choice(2) 
 0 Choice(1) 
 0 0 Leaf   10
 0 1 Leaf    6
 1 Choice(1) 
 1 0 Leaf    6
 1 1 Leaf   10
actual:
 f[ (1,2), (2,2), ]
 Choice(2) 
 0 Choice(1) 
 0 0 Leaf 0.3125
 0 1 Leaf 0.1875
 1 Choice(1) 
 1 0 Leaf 0.1875
 1 1 Leaf 0.3125
/home/phlee/gtsam/gtsam_scratch/gtsam/gtsam/discrete/tests/testDiscreteFactorGraph.cpp:122: Failure: "assert_equal(expectedFactor, *newFactor)" 
There were 1 failures
```

--- a/gtsam/discrete/DiscreteFactorGraph.cpp
+++ b/gtsam/discrete/DiscreteFactorGraph.cpp
@@ -210,6 +210,12 @@ namespace gtsam {
     for (auto&& factor : factors) product = (*factor) * product;
     gttoc(product);
 
+    // Sum all the potentials by pretending all keys are frontal:
+    auto normalization = product.sum(product.size());
+
+    // Normalize the product factor to prevent underflow.
+    product = product / (*normalization);
+
     // sum out frontals, this is the factor on the separator
     gttic(sum);
     DecisionTreeFactor::shared_ptr sum = product.sum(frontalKeys);

--- a/gtsam/discrete/DiscreteFactorGraph.cpp
+++ b/gtsam/discrete/DiscreteFactorGraph.cpp
@@ -121,8 +121,8 @@ namespace gtsam {
     for (auto&& factor : factors) product = (*factor) * product;
     gttoc(product);
 
-    // Sum all the potentials by pretending all keys are frontal:
-    auto normalization = product.sum(product.size());
+    // Max over all the potentials by pretending all keys are frontal:
+    auto normalization = product.max(product.size());
 
     // Normalize the product factor to prevent underflow.
     product = product / (*normalization);
@@ -210,8 +210,8 @@ namespace gtsam {
     for (auto&& factor : factors) product = (*factor) * product;
     gttoc(product);
 
-    // Sum all the potentials by pretending all keys are frontal:
-    auto normalization = product.sum(product.size());
+    // Max over all the potentials by pretending all keys are frontal:
+    auto normalization = product.max(product.size());
 
     // Normalize the product factor to prevent underflow.
     product = product / (*normalization);

--- a/gtsam/discrete/tests/testDiscreteFactorGraph.cpp
+++ b/gtsam/discrete/tests/testDiscreteFactorGraph.cpp
@@ -119,6 +119,9 @@ TEST(DiscreteFactorGraph, test) {
   // Check Factor
   CHECK(newFactor);
   DecisionTreeFactor expectedFactor(B & A, "10 6 6 10");
+  auto normalization = expectedFactor.sum(expectedFactor.size());
+  // Ensure normalization is correct.
+  expectedFactor = expectedFactor / *normalization;
   EXPECT(assert_equal(expectedFactor, *newFactor));
 
   // Test using elimination tree

--- a/gtsam/discrete/tests/testDiscreteFactorGraph.cpp
+++ b/gtsam/discrete/tests/testDiscreteFactorGraph.cpp
@@ -108,7 +108,14 @@ TEST(DiscreteFactorGraph, test) {
 
   // Test EliminateDiscrete
   const Ordering frontalKeys{0};
-  const auto [conditional, newFactor] = EliminateDiscrete(graph, frontalKeys);
+  const auto [conditional, newFactorPtr] = EliminateDiscrete(graph, frontalKeys);
+
+  DecisionTreeFactor newFactor = *newFactorPtr;
+
+  // Normalize newFactor by max for comparison with expected
+  auto normalization = newFactor.max(newFactor.size());
+
+  newFactor = newFactor / *normalization;
 
   // Check Conditional
   CHECK(conditional);
@@ -117,12 +124,13 @@ TEST(DiscreteFactorGraph, test) {
   EXPECT(assert_equal(expectedConditional, *conditional));
 
   // Check Factor
-  CHECK(newFactor);
+  CHECK(&newFactor);
   DecisionTreeFactor expectedFactor(B & A, "10 6 6 10");
-  auto normalization = expectedFactor.sum(expectedFactor.size());
+  // Normalize by max.
+  normalization = expectedFactor.max(expectedFactor.size());
   // Ensure normalization is correct.
   expectedFactor = expectedFactor / *normalization;
-  EXPECT(assert_equal(expectedFactor, *newFactor));
+  EXPECT(assert_equal(expectedFactor, newFactor));
 
   // Test using elimination tree
   const Ordering ordering{0, 1, 2};

--- a/python/gtsam/tests/test_DiscreteFactorGraph.py
+++ b/python/gtsam/tests/test_DiscreteFactorGraph.py
@@ -14,7 +14,7 @@ Author: Frank Dellaert
 import unittest
 
 import numpy as np
-from gtsam import DiscreteConditional, DiscreteFactorGraph, DiscreteKeys, DiscreteValues, Ordering, Symbol
+from gtsam import DecisionTreeFactor, DiscreteConditional, DiscreteFactorGraph, DiscreteKeys, DiscreteValues, Ordering, Symbol
 from gtsam.utils.test_case import GtsamTestCase
 
 OrderingType = Ordering.OrderingType
@@ -250,7 +250,7 @@ class TestDiscreteFactorGraph(GtsamTestCase):
 
         # Ensure that the stationary distribution is positive and normalized
         stationary_dist /= np.sum(stationary_dist)
-        expected = stationary_dist.flatten()
+        expected = DecisionTreeFactor(X[chain_length-1], stationary_dist.flatten())
 
         # The transition matrix parsed by DiscreteConditional is a row-wise CPT
         transitions = transitions.T
@@ -271,16 +271,8 @@ class TestDiscreteFactorGraph(GtsamTestCase):
         # Get the DiscreteConditional representing the marginal on the last factor
         last_marginal = sum_product.at(chain_length - 1)
 
-        # Extract the actual marginal probabilities
-        assignment = DiscreteValues()
-        marginal_probs = []
-        for i in range(num_states):
-            assignment[X[chain_length - 1][0]] = i
-            marginal_probs.append(last_marginal(assignment))
-        marginal_probs = np.array(marginal_probs)
-
         # Ensure marginal probabilities are close to the stationary distribution
-        self.gtsamAssertEquals(expected, marginal_probs)
+        self.gtsamAssertEquals(expected, last_marginal)
 
 if __name__ == "__main__":
     unittest.main()

--- a/python/gtsam/tests/test_DiscreteFactorGraph.py
+++ b/python/gtsam/tests/test_DiscreteFactorGraph.py
@@ -216,5 +216,70 @@ class TestDiscreteFactorGraph(GtsamTestCase):
 
         self.assertEqual(vals, [desired_state]*num_obs)
 
+    def test_sumProduct_chain(self):
+        """
+        Test for numerical underflow in EliminateDiscrete on long chains.
+        Adapted from the toy problem of @pcl15423
+        Ref: https://github.com/borglab/gtsam/issues/1448
+        """
+        num_states = 3
+        num_obs = 200
+        desired_state = 1
+        states = list(range(num_states))
+
+        # Helper function to mimic the behavior of gtbook.Variables discrete_series function
+        def make_key(character, index, cardinality):
+            symbol = Symbol(character, index)
+            key = symbol.key()
+            return (key, cardinality)
+
+        X = {index: make_key("X", index, len(states)) for index in range(num_obs)}
+        Z = {index: make_key("Z", index, num_obs + 1) for index in range(num_obs)}
+        graph = DiscreteFactorGraph()
+
+        # Mostly identity transition matrix
+        transitions = np.eye(num_states)
+
+        # Needed otherwise mpe is always state 0?
+        transitions += 0.1/(num_states)
+
+        transition_cpt = []
+        for i in range(0, num_states):
+            transition_row = "/".join([str(x) for x in transitions[i]])
+            transition_cpt.append(transition_row)
+        transition_cpt = " ".join(transition_cpt)
+
+        for i in reversed(range(1, num_obs)):
+            transition_conditional = DiscreteConditional(X[i], [X[i-1]], transition_cpt)
+            graph.push_back(transition_conditional)
+
+        # Contrived example such that the desired state gives measurements [0, num_obs) with equal probability
+        #   but all other states always give measurement num_obs
+        obs = np.zeros((num_states, num_obs+1))
+        obs[:,-1] = 1
+        obs[desired_state,0: -1] = 1
+        obs[desired_state,-1] = 0
+        obs_cpt_list = []
+        for i in range(0, num_states):
+            obs_row = "/".join([str(z) for z in obs[i]])
+            obs_cpt_list.append(obs_row)
+        obs_cpt = " ".join(obs_cpt_list)
+
+        # Contrived example where each measurement is its own index
+        for i in range(0, num_obs):
+            obs_conditional = DiscreteConditional(Z[i], [X[i]], obs_cpt)
+            factor = obs_conditional.likelihood(i)
+            graph.push_back(factor)
+
+        mpe = graph.optimize()
+        vals = [mpe[X[i][0]] for i in range(num_obs)]
+        sum_product = graph.sumProduct()
+
+        print("This should have 9 potential assignments", sum_product.at(0))
+
+        print("This should have 9 potential assignments", sum_product.at(138))
+
+        self.assertEqual(vals, [desired_state]*num_obs)
+
 if __name__ == "__main__":
     unittest.main()

--- a/python/gtsam/tests/test_DiscreteFactorGraph.py
+++ b/python/gtsam/tests/test_DiscreteFactorGraph.py
@@ -223,7 +223,7 @@ class TestDiscreteFactorGraph(GtsamTestCase):
         Ref: https://github.com/borglab/gtsam/issues/1448
         """
         num_states = 3
-        num_obs = 200
+        chain_length = 400
         desired_state = 1
         states = list(range(num_states))
 
@@ -233,53 +233,54 @@ class TestDiscreteFactorGraph(GtsamTestCase):
             key = symbol.key()
             return (key, cardinality)
 
-        X = {index: make_key("X", index, len(states)) for index in range(num_obs)}
-        Z = {index: make_key("Z", index, num_obs + 1) for index in range(num_obs)}
+        X = {index: make_key("X", index, len(states)) for index in range(chain_length)}
         graph = DiscreteFactorGraph()
 
-        # Mostly identity transition matrix
-        transitions = np.eye(num_states)
-
-        # Needed otherwise mpe is always state 0?
+        # Construct test transition matrix
+        transitions = np.diag([1.0, 0.5, 0.1])
         transitions += 0.1/(num_states)
 
+        # Ensure that the transition matrix is Markov (columns sum to 1)
+        transitions /= np.sum(transitions, axis=0)
+
+        # The stationary distribution is the eigenvector corresponding to eigenvalue 1
+        eigvals, eigvecs = np.linalg.eig(transitions)
+        stationary_idx = np.where(np.isclose(eigvals, 1.0))
+        stationary_dist = eigvecs[:, stationary_idx]
+
+        # Ensure that the stationary distribution is positive and normalized
+        stationary_dist /= np.sum(stationary_dist)
+        expected = stationary_dist.flatten()
+
+        # The transition matrix parsed by DiscreteConditional is a row-wise CPT
+        transitions = transitions.T
         transition_cpt = []
         for i in range(0, num_states):
             transition_row = "/".join([str(x) for x in transitions[i]])
             transition_cpt.append(transition_row)
         transition_cpt = " ".join(transition_cpt)
 
-        for i in reversed(range(1, num_obs)):
+        for i in reversed(range(1, chain_length)):
             transition_conditional = DiscreteConditional(X[i], [X[i-1]], transition_cpt)
             graph.push_back(transition_conditional)
 
-        # Contrived example such that the desired state gives measurements [0, num_obs) with equal probability
-        #   but all other states always give measurement num_obs
-        obs = np.zeros((num_states, num_obs+1))
-        obs[:,-1] = 1
-        obs[desired_state,0: -1] = 1
-        obs[desired_state,-1] = 0
-        obs_cpt_list = []
-        for i in range(0, num_states):
-            obs_row = "/".join([str(z) for z in obs[i]])
-            obs_cpt_list.append(obs_row)
-        obs_cpt = " ".join(obs_cpt_list)
+        # Run sum product using natural ordering so the resulting Bayes net has the form:
+        # X_0 <- X_1 <- ... <- X_n
+        sum_product = graph.sumProduct(OrderingType.NATURAL)
 
-        # Contrived example where each measurement is its own index
-        for i in range(0, num_obs):
-            obs_conditional = DiscreteConditional(Z[i], [X[i]], obs_cpt)
-            factor = obs_conditional.likelihood(i)
-            graph.push_back(factor)
+        # Get the DiscreteConditional representing the marginal on the last factor
+        last_marginal = sum_product.at(chain_length - 1)
 
-        mpe = graph.optimize()
-        vals = [mpe[X[i][0]] for i in range(num_obs)]
-        sum_product = graph.sumProduct()
+        # Extract the actual marginal probabilities
+        assignment = DiscreteValues()
+        marginal_probs = []
+        for i in range(num_states):
+            assignment[X[chain_length - 1][0]] = i
+            marginal_probs.append(last_marginal(assignment))
+        marginal_probs = np.array(marginal_probs)
 
-        print("This should have 9 potential assignments", sum_product.at(0))
-
-        print("This should have 9 potential assignments", sum_product.at(138))
-
-        self.assertEqual(vals, [desired_state]*num_obs)
+        # Ensure marginal probabilities are close to the stationary distribution
+        self.gtsamAssertEquals(expected, marginal_probs)
 
 if __name__ == "__main__":
     unittest.main()

